### PR TITLE
HHH-13085 : Use FetchMode.JOIN to load OneToOne associations that hav…

### PR DESCRIPTION
**Pull request by gbadner**
…e the same AssociationKey as the entity owner

https://hibernate.atlassian.net/browse/HHH-13085

SQL generated for the following tests show that one-to-one associations are being joined:
rg.hibernate.test.annotations.derivedidentities.bidirectional.OneToOneWithDerivedIdentityTest
org.hibernate.test.annotations.derivedidentities.e4.a.DerivedIdentitySimpleParentSimpleDepTest
org.hibernate.test.annotations.derivedidentities.e4.b.DerivedIdentitySimpleParentSimpleDepMapsIdTest
org.hibernate.test.annotations.derivedidentities.e5.a.DerivedIdentityIdClassParentSameIdTypeIdClassDepTest
org.hibernate.test.annotations.derivedidentities.e5.b.DerivedIdentityIdClassParentSameIdTypeEmbeddedIdDepTest
org.hibernate.test.annotations.derivedidentities.e5.c.ForeignGeneratorViaMapsIdTest
org.hibernate.test.annotations.derivedidentities.e6.a.DerivedIdentityEmbeddedIdParentSameIdTypeIdClassDepTest
org.hibernate.test.annotations.derivedidentities.e6.b.DerivedIdentityEmbeddedIdParentSameIdTypeEmbeddedIdDepTest
org.hibernate.test.annotations.manytoone.ManyToOneMapsIdFlushModeTest
org.hibernate.test.annotations.notfound.NotFoundOneToOneNonInsertableNonUpdateableTest
org.hibernate.test.annotations.onetoone.OneToOneTest
org.hibernate.test.annotations.onetoone.OptionalOneToOneMappedByTest
org.hibernate.test.annotations.onetoone.OptionalOneToOneMapsIdQueryTest
org.hibernate.test.annotations.onetoone.OptionalOneToOnePKJCQueryTest
org.hibernate.test.annotations.onetoone.OptionalOneToOnePKJCTest
org.hibernate.test.criteria.mapsid.MapsIdOneToOneSelectTest

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/test/loadplans/walking/LoggingAssociationVisitationStrategy.java [hunk 1]
```diff
@@ -23,17 +29,32 @@
 /**
  * @author Steve Ebersole
  */
-public class LoggingAssociationVisitationStrategy implements AssociationVisitationStrategy {
+public class LoggingAssociationVisitationStrategy extends FetchStyleLoadPlanBuildingAssociationVisitationStrategy {
 	private int depth = 1;
 
+	/**
+	 * Constructs a FetchStyleLoadPlanBuildingAssociationVisitationStrategy.
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/test/loadplans/walking/LoggingAssociationVisitationStrategy.java:36]: seems we are constructing a `LoggingAssociationVisitationStrategy` here, so the above Javadoc is a little bit confusing

File: hibernate-core/src/main/java/org/hibernate/loader/plan/build/internal/AbstractLoadPlanBuildingAssociationVisitationStrategy.java [hunk 2]
```diff
@@ -640,6 +641,34 @@ public boolean isDuplicateAssociationKey(AssociationKey associationKey) {
...
+		else if ( registeredFetchSource instanceof Fetch ) {
+			registeredOwner = ( (Fetch) registeredFetchSource ).getSource().resolveEntityReference();
+		}
+		else {
+			registeredOwner = registeredFetchSource.resolveEntityReference();
+		}
+		return isDuplicateAssociationKey( associationKey ) &&
+				( !currentEntityReferenceAssociationKey.equals( attributeDefinition.getAssociationKey() ) ||
+						( registeredOwner != null && !currentEntityPersister.equals( registeredOwner.getEntityPersister() ) ) ||
+						currentEntityPersister.equals( attributeDefinition.toEntityDefinition().getEntityPersister() ) );
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/loader/plan/build/internal/AbstractLoadPlanBuildingAssociationVisitationStrategy.java:669]: The above logic seems complicated and the format seems not helpful to grasp the overall structure. Maybe joining the first two lines is considerable on this:
```
return ( isDuplicateAssociationKey( associationKey ) && !currentEntityReferenceAssociationKey.equals( attributeDefinition.getAssociationKey() ) ||
						( registeredOwner != null && !currentEntityPersister.equals( registeredOwner.getEntityPersister() ) ) ||
						currentEntityPersister.equals( attributeDefinition.toEntityDefinition().getEntityPersister() ) );
```

File: hibernate-core/src/main/java/org/hibernate/persister/walking/spi/MetamodelGraphWalker.java [hunk 3]
```diff
@@ -326,4 +326,16 @@ protected void addAssociationKey(AssociationKey associationKey) {
...
+	protected boolean isDuplicateAttributeDefinition(AssociationAttributeDefinition attributeDefinition) {
+		if ( isDuplicateAssociationKey( attributeDefinition.getAssociationKey() ) ) {
+			if ( attributeDefinition.getAssociationNature() == AssociationAttributeDefinition.AssociationNature.ENTITY ) {
+				return strategy.isDuplicateAssociatedEntity( attributeDefinition );
+			}
+			else {
+				return true;
+			}
+		}
+		return false;
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/persister/walking/spi/MetamodelGraphWalker.java:339]: I think we can make the above logic a little bit more compact as
```
return !isDuplicateAssociationKey( attributeDefinition.getAssociationKey() ||
           attributeDefinition.getAssociationNature() != AssociationAttributeDefinition.AssociationNature.ENTITY ||
           strategy.isDuplicateAssociatedEntity( attributeDefinition );
```
Or a little bit less aggressive as
```
if ( isDuplicateAssociationKey( attributeDefinition.getAssociationKey() ) ) {
    return attributeDefinition.getAssociationNature() != AssociationAttributeDefinition.AssociationNature.ENTITY || strategy.isDuplicateAssociatedEntity( attributeDefinition );
}
else {
    return false;
}
```
Subjective and minor. Feel free to ignore this comment.

File: hibernate-core/src/main/java/org/hibernate/loader/plan/build/internal/AbstractLoadPlanBuildingAssociationVisitationStrategy.java (hunk 2 above)
Comment (gbadner) [reply hibernate-core/src/main/java/org/hibernate/loader/plan/build/internal/AbstractLoadPlanBuildingAssociationVisitationStrategy.java:669]: @NathanQingyangXu, I, personally, don't find your change to be more readable.

I will add some comments to describe the logic, and I will try to simplify it.

File: hibernate-core/src/main/java/org/hibernate/persister/walking/spi/MetamodelGraphWalker.java (hunk 3 above)
Comment (gbadner) [reply hibernate-core/src/main/java/org/hibernate/persister/walking/spi/MetamodelGraphWalker.java:339]: @NathanQingyangXu I will add some comments here as well, and try to simplify the logic.

**Comment by gbadner** [comment]
I am in the process of creating unit tests that assert that FetchMode.JOIN is properly being used.

**Comment by gbadner** [comment]
I'm going to close this PR for now. It only has a limited benefit, and, IMO, it's too complicated and has the potential to cause regressions.

The approach taken in https://github.com/hibernate/hibernate-orm/pull/3873 is preferable.

**Pull request closed without merging**